after udp restore: system call to 'sendto' report errno=EACCES Permission denied

I am experimenting with udp checkpoint feature. I have a udp client program and a udp server program. It seems to be OK by looking at the log. But when I tried to restore back to the checkpoint. The udp program failed to call `sendto`, with errno=EACCES (Permission denied). It seems the socket has lost 'SO_BROADCAST' attribute after restored.

My system configuration is as follows:
CentOS Linux release 7.1.1503 (Core)
CRIU 3.9

The following are the commands that I issued to create the problem.
criu dump -t 14455 --tcp-established --shell-job -vvvv -o dump.log && echo OK
date;criu restore -d -vvv --tcp-established --shell-job -o restore.log && echo OK

Code:
```c++
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <linux/types.h>
#include <sys/types.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <string.h>
#include <errno.h>

static int main_srv(int argc, char **argv)
{
    printf("main_srv\n");
    struct sockaddr_in cli;
    char request[400], reply[400];
    int clilen, n;
    int sk;
    sk = socket(AF_INET, SOCK_DGRAM, 0); 
    /* set reuse and non block for this socket */
    int on=1;
    setsockopt(sk, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on));
    setsockopt(sk, SOL_SOCKET, SO_BROADCAST, (char *)&on, sizeof(on));

    struct sockaddr_in servaddr; 
    memset(&servaddr, 0, sizeof(servaddr)); 
    servaddr.sin_family = AF_INET; 
    servaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    servaddr.sin_port = htons(10000); 
    
    if (bind(sk, (struct sockaddr *) &servaddr, sizeof(servaddr)) != 0)
    {
          perror("Can't bind socket");
          return -1;
    }
    printf("bind to 10000\n");

    for (;;) {
        clilen = sizeof(cli);
        if ( (n=recvfrom(sk, request, 400, 0, (struct sockaddr *)&cli, &clilen)) < 0 ) {
            perror("recv from server error");
        }
        
        if ( sendto(sk, request, 400, 0, (struct sockaddr *)&cli, sizeof(cli)) != 400 ) {
            perror("sendto from server error");
        }
    }
}

static int main_cl(int argc, char **argv)
{
    printf("main_cl\n");
    char request[400], reply[400];
    int n;
    struct sockaddr_in servaddr; 
    int sk = socket(AF_INET, SOCK_DGRAM, 0); 
    if (sk < 0) {
        perror("Can't create socket");
        return -1;
    }
    /* set reuse and non block for this socket */
    int on=1;
    setsockopt(sk, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on));
    setsockopt(sk, SOL_SOCKET, SO_BROADCAST, (char *)&on, sizeof(on));

    struct sockaddr_in targetaddr; 
    memset(&targetaddr, 0, sizeof(targetaddr)); 
    targetaddr.sin_family = AF_INET; 
    //targetaddr.sin_addr.s_addr = htonl(INADDR_ANY);
    targetaddr.sin_addr.s_addr = inet_addr("127.255.255.255");
    targetaddr.sin_port = htons(10000); 

    int val = 0;
    while (1) {
        val++;
        snprintf(request, 400, "%d", val);
        if ( sendto(sk, request, 400, 0, (struct sockaddr *)&targetaddr, sizeof(targetaddr)) != 400 ) {
            perror("sendto from client error");
        }
        
        if ( (n=recvfrom(sk, reply, 400, 0, (struct sockaddr *)NULL, (int*) NULL)) < 0 ) {
            perror("recv from client error");
        }

        printf("UDP %s -> %s\n", request, reply);
        sleep(2);
    }
    
}

int main(int argc, char **argv)
{
    if (strcmp(argv[1], "server")==0 )
        return main_srv(argc, argv);
    else if (strcmp(argv[1],"client")==0 )
        return main_cl(argc, argv);

    printf("Bad usage\n");
    return 1;
}
```

Thanks! But now I came cross compilation problem. 

My system configuration is as follows:
CentOS Linux release 7.1.1503 (Core)
CRIU source code
gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC)

Build output after I follow the notes from  https://criu.org/Installation
```bash
make -r -R -f /home/guoqd/workspace/criu/criu/scripts/nmk/scripts/main.mk makefile=Makefile obj=soccr all
cc -c -O2 -g -Wall -Wformat-security -Werror -DCONFIG_X86_64 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -DCONFIG_HAS_SELINUX -DCONFIG_COMPAT -iquote include/ -DCONFIG_HAS_SELINUX -DCONFIG_COMPAT soccr/soccr.c -o soccr/soccr.o
soccr/soccr.c: 在函数‘tcp_repair_on’中:
soccr/soccr.c:85:32: 错误：‘TCP_REPAIR’未声明(在此函数内第一次使用)
  ret = setsockopt(fd, SOL_TCP, TCP_REPAIR, &aux, sizeof(aux));
                                ^
soccr/soccr.c:85:32: 附注：每个未声明的标识符在其出现的函数内只报告一次
soccr/soccr.c: 在函数‘tcp_repair_off’中:
soccr/soccr.c:96:32: 错误：‘TCP_REPAIR’未声明(在此函数内第一次使用)
  ret = setsockopt(fd, SOL_TCP, TCP_REPAIR, &aux, sizeof(aux));
                                ^
soccr/soccr.c: 在函数‘get_queue’中:
soccr/soccr.c:333:32: 错误：‘TCP_REPAIR_QUEUE’未声明(在此函数内第一次使用)
  ret = setsockopt(sk, SOL_TCP, TCP_REPAIR_QUEUE, &aux, auxl);
                                ^
soccr/soccr.c:338:32: 错误：‘TCP_QUEUE_SEQ’未声明(在此函数内第一次使用)
  ret = getsockopt(sk, SOL_TCP, TCP_QUEUE_SEQ, seq, &auxl);
                                ^
soccr/soccr.c: 在函数‘set_queue_seq’中:
soccr/soccr.c:449:34: 错误：‘TCP_REPAIR_QUEUE’未声明(在此函数内第一次使用)
  if (setsockopt(sk->fd, SOL_TCP, TCP_REPAIR_QUEUE, &queue, sizeof(queue)) < 0) {
                                  ^
soccr/soccr.c:454:34: 错误：‘TCP_QUEUE_SEQ’未声明(在此函数内第一次使用)
  if (setsockopt(sk->fd, SOL_TCP, TCP_QUEUE_SEQ, &seq, sizeof(seq)) < 0) {
                                  ^
soccr/soccr.c: 在函数‘libsoccr_set_sk_data_noq’中:
soccr/soccr.c:566:34: 错误：‘TCP_REPAIR_OPTIONS’未声明(在此函数内第一次使用)
      setsockopt(sk->fd, SOL_TCP, TCP_REPAIR_OPTIONS,
                                  ^
soccr/soccr.c: 在函数‘restore_fin_in_snd_queue’中:
soccr/soccr.c:710:30: 错误：‘TCP_REPAIR_QUEUE’未声明(在此函数内第一次使用)
      setsockopt(sk, SOL_TCP, TCP_REPAIR_QUEUE, &queue, sizeof(queue)) < 0) {
                              ^
soccr/soccr.c: 在函数‘send_queue’中:
soccr/soccr.c:847:34: 错误：‘TCP_REPAIR_QUEUE’未声明(在此函数内第一次使用)
  if (setsockopt(sk->fd, SOL_TCP, TCP_REPAIR_QUEUE, &queue, sizeof(queue)) < 0) {
                                  ^
make[1]: *** [soccr/soccr.o] 错误 1
make: *** [soccr/built-in.o] Error 2
```

--- a/criu/sk-inet.c
+++ b/criu/sk-inet.c
@@ -742,6 +742,10 @@ static int post_open_inet_sk(struct file_desc *d, int sk)
 	if (!val && restore_opt(sk, SOL_SOCKET, SO_REUSEPORT, &val))
 		return -1;
 
+	val = ii->ie->opts->so_broadcast;
+	if (!val && restore_opt(sk, SOL_SOCKET, SO_BROADCAST, &val))
+		return -1;
+
 	return 0;
 }
 

--- a/criu/sockets.c
+++ b/criu/sockets.c
@@ -566,6 +566,11 @@ int restore_socket_opts(int sk, SkOptsEntry *soe)
 		pr_debug("\tset no_check for socket\n");
 		ret |= restore_opt(sk, SOL_SOCKET, SO_NO_CHECK, &val);
 	}
+	if (soe->has_so_broadcast && soe->so_broadcast) {
+		val = 1;
+		pr_debug("\tset broadcast for socket\n");
+		ret |= restore_opt(sk, SOL_SOCKET, SO_BROADCAST, &val);
+	}
 
 	tv.tv_sec = soe->so_snd_tmo_sec;
 	tv.tv_usec = soe->so_snd_tmo_usec;
@@ -646,6 +651,10 @@ int dump_socket_opts(int sk, SkOptsEntry *soe)
 	ret |= dump_opt(sk, SOL_SOCKET, SO_NO_CHECK, &val);
 	soe->has_so_no_check = true;
 	soe->so_no_check = val ? true : false;
+
+	ret |= dump_opt(sk, SOL_SOCKET, SO_BROADCAST, &val);
+	soe->has_so_broadcast = true;
+	soe->so_broadcast = val ? true : false;
 
 	ret |= dump_bound_dev(sk, soe);
 	ret |= dump_socket_filter(sk, soe);

--- a/images/sk-opts.proto
+++ b/images/sk-opts.proto
@@ -22,6 +22,7 @@ message sk_opts_entry {
 
 	repeated fixed64	so_filter	= 16;
 	optional bool		so_reuseport	= 17;
+	optional bool		so_broadcast	= 18;
 }
 
 enum sk_shutdown {

--- a/test/zdtm/static/Makefile
+++ b/test/zdtm/static/Makefile
@@ -30,6 +30,7 @@ TST_NOFILE	:=				\
 		socket_listen6			\
 		socket_listen4v6		\
 		socket_udp			\
+		socket_udp-broadcast		\
 		socket_udp-corked		\
 		socket6_udp			\
 		socket_udp_shutdown		\

--- a/test/zdtm/static/socket_udp-broadcast.c
+++ b/test/zdtm/static/socket_udp-broadcast.c
@@ -1,0 +1,47 @@
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "test checkpoint/restore of SO_BROADCAST\n";
+const char *test_author = "Radostin Stoyanov <rstoyanov1@gmail.com>\n";
+
+/* Description:
+ * Create UDP socket, set SO_BROADCAST and verify its value after restore.
+ */
+
+int main(int argc, char **argv)
+{
+	int sockfd;
+	int val;
+	socklen_t len = sizeof(val);
+
+	test_init(argc, argv);
+
+	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sockfd < 0) {
+		pr_perror("Can't create socket");
+		return 1;
+	}
+
+	if (setsockopt(sockfd, SOL_SOCKET, SO_BROADCAST, &(int){ 1 }, len)) {
+		pr_perror("setsockopt");
+		return 1;
+	}
+
+	test_daemon();
+	test_waitsig();
+
+	if (getsockopt(sockfd, SOL_SOCKET, SO_BROADCAST, &val, &len)) {
+		pr_perror("getsockopt");
+		return 1;
+	}
+
+	if (len != sizeof(val) || val != 1) {
+		fail("SO_BROADCAST not set");
+		return 1;
+	}
+
+	pass();
+	return 0;
+}

